<a href="https://colab.research.google.com/github/jbell1991/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module3-join-and-reshape-data/Copy_of_LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [45]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-12-04 23:08:09--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.169.37
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.169.37|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  35.4MB/s    in 6.1s    

2019-12-04 23:08:16 (32.1 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [46]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [47]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01/instacart_2017_05_01


In [48]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
##### YOUR CODE HERE #####
import pandas as pd
# We need the products, order products, and order products train dataframes

In [50]:
# Looking at products we need the product_id and product_name columns
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [51]:
# Looking at order_products_prior we need product_id and order_id
order_products__prior = pd.read_csv('order_products__prior.csv')

print(order_products__prior.shape)
order_products__prior.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [52]:
# Looking at order_products_train we need product_id and order_id
order_products__train = pd.read_csv('order_products__train.csv')

print(order_products__train.shape)
order_products__train.head()

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [53]:
# Joining the order_prodcuts_train and order_prodcuts prior together using concat
orders = pd.concat([order_products__train, order_products__prior])

print(orders.shape)
orders.head()

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [54]:
# Creating a dataframe containing each of the 10 most frequently ordered products
condition = (products['product_name'] == 'Banana') | (
             products['product_name'] == 'Bag of Organic Bananas') | (
             products['product_name'] == 'Organic Strawberries') | (
             products['product_name'] == 'Organic Baby Spinach') | (
             products['product_name'] == 'Organic Hass Avocado') | (
             products['product_name'] == 'Organic Avocado') | (
             products['product_name'] == 'Large Lemon') | (
             products['product_name'] == 'Strawberries') | (
             products['product_name'] == 'Limes') | (
             products['product_name'] == 'Organic Whole Milk')

products_subset = products[condition]

print(products_subset.shape)
products_subset.head(10)

(10, 4)


product_id            product_name  aisle_id  department_id
13175       13176  Bag of Organic Bananas        24              4
16796       16797            Strawberries        24              4
21136       21137    Organic Strawberries        24              4
21902       21903    Organic Baby Spinach       123              4
24851       24852                  Banana        24              4
26208       26209                   Limes        24              4
27844       27845      Organic Whole Milk        84             16
47208       47209    Organic Hass Avocado        24              4
47625       47626             Large Lemon        24              4
47765       47766         Organic Avocado        24              4

In [55]:
# Creating a dataframe containing the product_ids for the 10 most frequently 
# ordered products 
condition = (orders['product_id'] == 13176) | (
             orders['product_id'] == 16797) | (
             orders['product_id'] == 21137) | (
             orders['product_id'] == 21903) | (
             orders['product_id'] == 24852) | (
             orders['product_id'] == 26209) | (
             orders['product_id'] == 27845) | (
             orders['product_id'] == 47209) | (
             orders['product_id'] == 47626) | (
             orders['product_id'] == 47766)

orders_subset = orders[condition]

print(orders_subset.shape)
orders_subset.head(10)


(2418314, 4)


order_id  product_id  add_to_cart_order  reordered
5           1       13176                  6          0
6           1       47209                  7          0
35         98       13176                  4          1
85        112       47209                  5          0
94        170       13176                  3          1
104       170       21137                 13          0
115       226       24852                  2          0
143       393       16797                  6          1
154       456       47626                  9          0
156       473       24852                  2          0

In [56]:
# Merging the products_subset and orders_subset dataframes together by their unique identifier,
# product_id
orders_and_products = pd.merge(products_subset,
                               orders_subset,
                               on='product_id',
                               how='inner')

orders_and_products.head()

product_id            product_name  ...  add_to_cart_order  reordered
0       13176  Bag of Organic Bananas  ...                  6          0
1       13176  Bag of Organic Bananas  ...                  4          1
2       13176  Bag of Organic Bananas  ...                  3          1
3       13176  Bag of Organic Bananas  ...                  1          1
4       13176  Bag of Organic Bananas  ...                  2          1

[5 rows x 7 columns]

In [0]:
# Dropping the unnecessary columns 
orders_and_products = orders_and_products.drop(['aisle_id', 'department_id', 'add_to_cart_order', 'reordered'], axis=1)

In [71]:
# Checking to see if those columns have been dropped in a different cell to
# avoid rerunning the cell with the .drop method 
orders_and_products.head()

product_id            product_name  order_id
0       13176  Bag of Organic Bananas         1
1       13176  Bag of Organic Bananas        98
2       13176  Bag of Organic Bananas       170
3       13176  Bag of Organic Bananas      1077
4       13176  Bag of Organic Bananas      1325

In [69]:
# Checking to see the order counts of each product name and that the order matches
# the order given in the question
orders_and_products['product_name'].value_counts()

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [72]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
# Replicating the lesson code
# Take the row index, and add it as a new column
table1 =table1.reset_index()

In [74]:
table1

index  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

In [0]:
# What is the unique indentifier for each row
# Where is the data at that I want to be in my single 'tidy' column

tidy1 = table1.melt(id_vars='index', value_vars=['treatmenta', 'treatmentb'])

In [79]:
tidy1

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [80]:
# rename columns

tidy1 = tidy1.rename(columns={
    'index': 'name',
    'variable': 'trt',
    'value': 'result'
})
tidy1

name         trt  result
0    John Smith  treatmenta     NaN
1      Jane Doe  treatmenta    16.0
2  Mary Johnson  treatmenta     3.0
3    John Smith  treatmentb     2.0
4      Jane Doe  treatmentb    11.0
5  Mary Johnson  treatmentb     1.0

In [82]:
tidy1.trt = tidy1.trt.str.replace('treatment', '')

tidy1

name trt  result
0    John Smith   a     NaN
1      Jane Doe   a    16.0
2  Mary Johnson   a     3.0
3    John Smith   b     2.0
4      Jane Doe   b    11.0
5  Mary Johnson   b     1.0

In [83]:
# index: unique identifier
# columns: What do you want to differentitate the columns in wide format
# values: Where are the numbers at - go in the middle of the wide dataframe

wide = tidy1.pivot_table(index='name', columns='trt', values='result')
wide

trt              a     b
name                    
Jane Doe      16.0  11.0
John Smith     NaN   2.0
Mary Johnson   3.0   1.0

In [87]:
# table2 -> tidy
# Take the row index, and add it as a new column
table2 =table2.reset_index()

table2

level_0       index  John Smith  Jane Doe  Mary Johnson
0        0  treatmenta         NaN      16.0           3.0
1        1  treatmentb         2.0      11.0           1.0

In [94]:
# What is the unique identifier for each row
# Where is the data at that I want to be in my single "tidy" column
tidy2 = table2.melt(id_vars='index', value_vars=['John Smith', 'Jane Doe', 'Mary Johnson'])

tidy2

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [95]:
# rename columns 

tidy2 = tidy2.rename(columns={
    'variable': 'name',
    'index': 'trt',
    'value': 'result'
})
tidy2

trt          name  result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [98]:
# reorder columns
tidy2 = tidy2[['name', 'trt', 'result']]
tidy2

name         trt  result
0    John Smith  treatmenta     NaN
1    John Smith  treatmentb     2.0
2      Jane Doe  treatmenta    16.0
3      Jane Doe  treatmentb    11.0
4  Mary Johnson  treatmenta     3.0
5  Mary Johnson  treatmentb     1.0

In [100]:
# tidy -> table2
wide = tidy2.pivot_table(index='trt', columns='name', values='result')
wide

name        Jane Doe  John Smith  Mary Johnson
trt                                           
treatmenta      16.0         NaN           3.0
treatmentb      11.0         2.0           1.0

In [103]:
# Needed to import seaborn 
import seaborn as sns
flights = sns.load_dataset('flights')
flights.head()

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [104]:
# Flights Pivot Table

flights_pivot_table = flights.pivot_table(index='year', columns='month', values = 'passengers')
flights_pivot_table 

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####